### PR TITLE
Issue/4019 attributes starting with a capital letter could get better exceptions

### LINKED ISSUE #4019: Attributes starting with a capital letter could get better exceptions
This simple model will raise a parser exception because the attribute doesn't start with a lower case letter, but the error doesn't mention this is actually the issue.
```
entity A:
    bool CRC16
end
```

The error raised is:
```
Syntax error at token CRC16 (./main.cf:2:15)
```

<details>
<summary>Full trace</summary>

```
Traceback (most recent call last):
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/app.py", line 780, in app
    options.func(options)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/app.py", line 355, in compile_project
    do_compile()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 69, in do_compile
    compiler.handle_exception(e)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 248, in handle_exception
    self._handle_exception_export(e)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 254, in _handle_exception_export
    raise exception
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 246, in handle_exception
    self._handle_exception_datatrace(exception)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 258, in _handle_exception_datatrace
    raise exception
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 67, in do_compile
    (statements, blocks) = compiler.compile()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 188, in compile
    project.load()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1616, in load
    self.get_complete_ast()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1651, in get_complete_ast
    (statements, block) = self.get_ast()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1636, in get_ast
    return self.__load_ast()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1666, in __load_ast
    return self._load_file(main_ns, os.path.join(self.project_path, self.main_file))
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1334, in _load_file
    stmts = plyInmantaParser.parse(ns, file)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/parser/plyInmantaParser.py", line 1172, in parse
    statements = base_parse(namespace, filename, content)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/parser/plyInmantaParser.py", line 1153, in base_parse
    return parser.parse(data, lexer=lexer, debug=False)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/ply/yacc.py", line 1201, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/ply/yacc.py", line 192, in call_errorfunc
    r = errorfunc(token)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/parser/plyInmantaParser.py", line 1126, in p_error
    raise ParserException(r, p.value)
inmanta.parser.ParserException: Syntax error at token CRC16 (./main.cf:2:15)
```

</details>

When the type of the attribute is a dict it gets much worse (not sure if this is related or not, it might require another ticket).  
This model
```
entity A:
    dict CRC16
end
```
raises the exception
```
Syntax error: invalid identifier, dict is a reserved keyword (./main.cf:2:5)
```

<details>
<summary>Full trace</summary>

```
Traceback (most recent call last):
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/app.py", line 780, in app
    options.func(options)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/app.py", line 355, in compile_project
    do_compile()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 69, in do_compile
    compiler.handle_exception(e)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 248, in handle_exception
    self._handle_exception_export(e)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 254, in _handle_exception_export
    raise exception
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 246, in handle_exception
    self._handle_exception_datatrace(exception)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 258, in _handle_exception_datatrace
    raise exception
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 67, in do_compile
    (statements, blocks) = compiler.compile()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/compiler/__init__.py", line 188, in compile
    project.load()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1616, in load
    self.get_complete_ast()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1651, in get_complete_ast
    (statements, block) = self.get_ast()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1636, in get_ast
    return self.__load_ast()
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1666, in __load_ast
    return self._load_file(main_ns, os.path.join(self.project_path, self.main_file))
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/module.py", line 1334, in _load_file
    stmts = plyInmantaParser.parse(ns, file)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/parser/plyInmantaParser.py", line 1172, in parse
    statements = base_parse(namespace, filename, content)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/parser/plyInmantaParser.py", line 1153, in base_parse
    return parser.parse(data, lexer=lexer, debug=False)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/ply/yacc.py", line 333, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/ply/yacc.py", line 1201, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/ply/yacc.py", line 192, in call_errorfunc
    r = errorfunc(token)
  File "/home/guillaume/.virtualenvs/yang-module-generator/lib/python3.9/site-packages/inmanta/parser/plyInmantaParser.py", line 1122, in p_error
    raise ParserException(
inmanta.parser.ParserException: Syntax error: invalid identifier, dict is a reserved keyword (./main.cf:2:5)
```

</details>

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description
An exception is now raised when an attribute is defined with a name that starts with a capital letter

closes #4019 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
~~- [ ] Type annotations are present~~
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/parser/plyInmantaParser.py [hunk 1]
```diff
@@ -368,6 +376,20 @@ def p_attr_undef(p: YaccProduction) -> None:
     attach_from_string(p, 2)
 
 
+def p_attr_dict_err(p: YaccProduction) -> None:
+    """attr : DICT CID empty
+    | DICT CID '=' map_def
+    | DICT CID '=' NULL
```
Comment (sanderr) [inline src/inmanta/parser/plyInmantaParser.py:382]:
```suggestion
    | DICT empty CID '=' map_def
    | DICT empty CID '=' NULL
```
I think this should allow you to use `p[3]` for all scenarios.

File: src/inmanta/parser/plyInmantaParser.py [hunk 2]
```diff
@@ -349,6 +349,14 @@ def p_attribute_type(p: YaccProduction) -> None:
     p[0] = p[1]
 
 
+def p_attr_err(p: YaccProduction) -> None:
+    """attr : attr_type CID empty
+    | attr_type CID '=' constant
+    | attr_type CID '=' constant_list
+    | attr_type CID '=' UNDEF"""
+    raise ParserException(p[2].location, str(p[2]), "Invalid identifier: Variable names must start with a lower case character")
```
Comment (sanderr) [inline src/inmanta/parser/plyInmantaParser.py:357]:
```suggestion
    raise ParserException(p[2].location, str(p[2]), "Invalid identifier: attribute names must start with a lower case character")
```

File: src/inmanta/parser/plyInmantaParser.py [hunk 3]
```diff
@@ -368,6 +376,20 @@ def p_attr_undef(p: YaccProduction) -> None:
...
+def p_attr_dict_err(p: YaccProduction) -> None:
+    """attr : DICT CID empty
+    | DICT CID '=' map_def
+    | DICT CID '=' NULL
+    | DICT '?' CID empty
+    | DICT '?'  CID '=' map_def
+    | DICT '?'  CID '=' NULL"""
+    if p[2] == "?":
+        raise ParserException(
+            p[3].location, str(p[3]), "Invalid identifier: Variable names must start with a lower case character"
```
Comment (sanderr) [inline src/inmanta/parser/plyInmantaParser.py:388]:
```suggestion
            p[3].location, str(p[3]), "Invalid identifier: attribute names must start with a lower case character"
```

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/parser/plyInmantaParser.py (hunk 1 above)
Comment (Hugo-Inmanta) [reply src/inmanta/parser/plyInmantaParser.py:382]: Neat trick

**Review by sanderr — APPROVED** [review]

File: changelogs/unreleased/4019-attributes-starting-with-a-capital-letter-could-get-better-exceptions.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+description: Improve syntax error reporting when defining an attribute starting with a capital letter.
+change-type: patch
+destination-branches: [master, iso5, iso4]
+sections: {
+}
```
Comment (arnaudsjs) [inline changelogs/unreleased/4019-attributes-starting-with-a-capital-letter-could-get-better-exceptions.yml:5]: Why not add this to the `bugfix` section?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Comment by Hugo-Inmanta** [comment]
Successfully merged into master (512154065fa4e358caf77bdf1bd0ebb92f4ead7c) and iso4 (44f9ccfd76e0fd9cc221dd4bf68758c9cf23bfab).

**Pull request closed without merging**